vmm/src/device_manager.rs

```diff
@@ -3123,6 +3123,16 @@ impl DeviceManager {
             })
             .map_err(DeviceManagerError::VfioUserMapRegion)?;
 
+        let memory = self.memory_manager.lock().unwrap().guest_memory();
+        let vfio_user_mapping = Arc::new(VfioUserDmaMapping::new(client, Arc::new(memory)));
+        for virtio_mem_device in self.virtio_mem_devices.iter() {
+            virtio_mem_device
+                .lock()
+                .unwrap()
+                .add_dma_mapping_handler(vfio_user_mapping.clone())
+                .map_err(DeviceManagerError::AddDmaMappingHandlerVirtioMem)?;
+        }
```

Comment:
I think you need to handle the unplug use case. If at any point in time the user unplugs the vfio-user device, the handler will become invalid. Therefore you should remove the handler from each `virtio-mem` device. But I realize that vfio-user behaves differently compared to VFIO, by creating one handler per device, where VFIO only needs one handler (since everything is under the same VFIO container).
That means you need to revert the change I applied [here](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3123/commits/1fbfbf4cf39a9ebc6866513e8b03a5d49f4fb391#diff-c67a1cf0a7a11141afda937bffab7be9b31d0f0de2660f064572f2a430fc716dL845-L890), and then you must provide a device_id when adding the handler, so that you can remove it later from the function `eject_device()`.